go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	}
+
+	for _, p := range w.pages {
+		defer p.Release()
+		w.WriteDataPage(p)
+	}
+	w.pages = w.pages[:0]
+	w.totalCompressedBytes = 0
+}
+
+func (w *columnWriter) writeLevels(numValues int64, defLevels, repLevels []int16) int64 {
+	toWrite := int64(0)
+	// if the field is required and non-repeated, no definition levels
+	if defLevels != nil && w.descr.MaxDefinitionLevel() > 0 {
+		for _, v := range defLevels {
+			if v == w.descr.MaxDefinitionLevel() {
+				toWrite++
+			}
+		}
+		w.WriteDefinitionLevels(defLevels[:numValues])
+	} else {
+		toWrite = numValues
+	}
+
+	if repLevels != nil && w.descr.MaxRepetitionLevel() > 0 {
+		// a row could include more than one value
+		//count the occasions where we start a new row
+		for _, v := range repLevels {
+			if v == 0 {
+				w.rowsWritten++
+			}
+		}
+
+		w.WriteRepetitionLevels(repLevels[:numValues])
+	} else {
+		// each value is exactly 1 row
+		w.rowsWritten += int(numValues)
+	}
+	return toWrite
+}
+
+func (w *columnWriter) writeLevelsSpaced(numLevels int64, defLevels, repLevels []int16) {
+	if w.descr.MaxDefinitionLevel() > 0 {
+		w.WriteDefinitionLevels(defLevels[:numLevels])
+	}
+
+	if w.descr.MaxRepetitionLevel() > 0 {
+		for _, v := range repLevels {
+			if v == 0 {
+				w.rowsWritten++
+			}
+		}
+		w.WriteRepetitionLevels(repLevels[:numLevels])
+	} else {
+		w.rowsWritten += int(numLevels)
+	}
+}
+
+func (w *columnWriter) WriteDictionaryPage() error {
+	dictEncoder := w.currentEncoder.(encoding.DictEncoder)
+	buffer := memory.NewResizableBuffer(w.mem)
+	buffer.Resize(dictEncoder.DictEncodedSize())
+	dictEncoder.WriteDict(buffer.Bytes())
+	defer buffer.Release()
+
+	page := NewDictionaryPage(buffer, int32(dictEncoder.NumEntries()), w.props.DictionaryPageEncoding())
+	written, err := w.pager.WriteDictionaryPage(page)
+	w.totalBytesWritten += written
+	return err
+}
+
+// this will always update the three otuput params
+// outValsToWrite, outSpacedValsToWrite, and NullCount. Additionally
+// it will update the validity bitmap if required (i.e. if at least one
+// level of nullable structs directly preced the leaf node)
+func (w *columnWriter) maybeCalculateValidityBits(defLevels []int16, batchSize int64, valuesToWrite, spacedValuesToWrite, nullCount *int64) {
+	if w.bitsBuffer == nil {
+		if w.levelInfo.DefLevel == 0 {
+			// in this case def levels should be null and we only
+			// need to output counts which will always be equal to
+			// the batch size passed in (max def level == 0 indicates
+			// there cannot be repeated or null fields)
+			*valuesToWrite = batchSize
+			*spacedValuesToWrite = batchSize
+			*nullCount = 0
+		} else {
+			var (
+				toWrite       int64
+				spacedToWrite int64
+			)
+			for i := int64(0); i < batchSize; i++ {
+				if defLevels[i] == w.levelInfo.DefLevel {
+					toWrite++
+				}
+				if defLevels[i] >= w.levelInfo.RepeatedAncestorDefLevel {
+					spacedToWrite++
+				}
+			}
+			*valuesToWrite += toWrite
+			*spacedValuesToWrite += spacedToWrite
```

Comment:
is spacedValuesToWrite needed would just returning valuesToWrite and nullcount be sufficient?

Reply:
technically you're correct (and i've modified the code as such) having them as three separate values did enable a couple fewer subtractions/additions but in theory the compiler should be able to recognize and optimize. I haven't benchmarked it yet but I wouldn't expect this to be an issue. If future benchmarking shows it as a potential performance improvement then I'll modify it at that point. for now the simplified code is better.